lib/vector-core/src/event/value.rs

```diff
@@ -22,6 +23,41 @@ pub enum Value {
     Null,
 }
 
+impl Eq for Value {}
+
+impl Hash for Value {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            Value::Array(v) => {
+                v.hash(state);
+            }
+            Value::Boolean(v) => {
+                v.hash(state);
+            }
+            Value::Bytes(v) => {
+                v.hash(state);
+            }
+            Value::Float(v) => {
+                if v.is_finite() {
+                    format!("{}", v).hash(state);
+                } else {
+                    "NaN".hash(state)
+                }
+            }
```

Comment:
This implementation feels pretty expensive in that it requires serialisation.

Hashing `f64`s, in general, feels pretty error prone, as discussed on https://internals.rust-lang.org/t/f32-f64-should-implement-hash/5436/33 . Seemingly the last suggestion there could work for us though:

* Have `NaN`s hash as equal. You are doing that here, but I think we could do it differently than hashing a string. 
* Ensure that `-0` and `+0` do not hash to the same value
* Otherwise transmute to u64 and hash

Reply:
@jszwedko What do you think about not allowing floats as a key? Is there a good way to prevent that behavior in vector?

Reply:
@dbcfd I think we should allow them for consistency; the alternative would be to error at runtime when an event comes though with a float for the value of the field being reduced. I think the comparison I outlined above should be sufficient though.